cupyx/scipy/ndimage/filters.py

```diff
@@ -0,0 +1,191 @@
+    The array is convolved with the given kernel.
+
+    Args:
+        input (cupy.ndarray): The input array.
+        weights (cupy.ndarray): Array of weights, same number of dimensions as
+            input
+        output (cupy.ndarray or None): The array in which to place the output.
+        mode (str): The array borders are handled according to the given mode
+            (``'reflect'``, ``'constant'``, ``'nearest'``, ``'mirror'``,
+            ``'wrap'``). Default is ``'reflect'``.
+        cval (scalar): Value to fill past edges of input if mode is
+            ``constant``. Default is ``0.0``.
+        origin (scalar or tuple of scalar): The origin parameter controls the
+            placement of the filter, relative to the center of the current
+            element of the input. Default of 0 is equivalent to
+            ``(0,)*input.ndim``.
+
+    Returns:
+        cupy.ndarray: The result of convolution.
+
+    .. seealso:: :func:`scipy.ndimage.convolve`
+    """
+    return _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                                  True)
+
+
+def _get_output(output, input, shape=None):
+    if shape is None:
+        shape = input.shape
+    if isinstance(output, cupy.ndarray):
+        if output.shape != tuple(shape):
+            raise ValueError('output shape is not correct')
+    else:
+        dtype = output
+        if dtype is None:
+            dtype = input.dtype
+        output = cupy.zeros(shape, dtype)
+    return output
+
+
+def _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                           convolution):
+    if input.dtype in (numpy.complex64, numpy.complex128, numpy.complex256):
+        raise TypeError('Complex type not supported.')
+    if not hasattr(origin, '__getitem__'):
+        origin = [origin, ] * input.ndim
+    wshape = [ii for ii in weights.shape if ii > 0]
+    if len(wshape) != input.ndim:
+        raise RuntimeError('filter weights array has incorrect shape.')
+    if convolution:
+        weights = weights[tuple([slice(None, None, -1)] * weights.ndim)]
+        for ii in range(len(origin)):
+            origin[ii] = - origin[ii]
+            if not weights.shape[ii] & 1:
+                origin[ii] -= 1
+    for _origin, lenw in zip(origin, wshape):
+        if (lenw // 2 + _origin < 0) or (lenw // 2 + _origin > lenw):
+            raise ValueError('invalid origin')
+    if mode not in ('reflect', 'constant', 'nearest', 'mirror', 'wrap'):
+        msg = 'boundary mode not supported (actual: {}).'.format(mode)
+        raise RuntimeError(msg)
+
+    output = _get_output(output, input)
+    input = cupy.ascontiguousarray(input)
+    weights = cupy.ascontiguousarray(weights, cupy.float64)
+    in_params, out_params, operation, name = _generate_correlete_kernel(
+        input.ndim, mode, cval, input.shape, wshape, origin)
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)(
+        input, weights, output)
```

Comment:
You don't have to fix in this PR but just FYI: this call will cause device synchronization due to a bug #2469.